lib/src/config_cache.dart

```diff
@@ -89,23 +86,57 @@ final class ConfigCache {
   ///     formatter:
   ///       page_width: 123
   Future<int?> findPageWidth(File file) async {
-    // Use the cached version (which may be `null`) if present.
+    return (await _findFormatterOptions(file)).pageWidth;
+  }
+
+  /// Looks for an "analysis_options.yaml" file surrounding [file] and, if
+  /// found and valid, returns the trailing comma handling specified by that
+  /// config file.
+  ///
+  /// Otherwise returns `null`.
+  ///
+  /// The schema looks like:
+  ///
+  ///     formatter:
+  ///       trailing_commas: preserve # Or "automate".
+  Future<TrailingCommas?> findTrailingCommas(File file) async {
+    return (await _findFormatterOptions(file)).trailingCommas;
+  }
+
+  /// Looks for an "analysis_options.yaml" file surrounding [file] and, if
+  /// found and valid, returns the configured options.
+  ///
+  /// If no options file could be found or it doesn't contain a "formatter" key
+  /// whose value is a map, returns a default set of options where all settings
+  /// are `null`.
+  Future<_FormatterOptions> _findFormatterOptions(File file) async {
+    // Use the cached version if present.
     var directory = file.parent.path;
-    if (_directoryPageWidths.containsKey(directory)) {
-      return _directoryPageWidths[directory];
-    }
+    if (_directoryOptions[directory] case var options?) return options;
+
+    int? pageWidth;
+    TrailingCommas? trailingCommas;
 
     try {
-      // Look for a surrounding analysis_options.yaml file.
-      var options = await findAnalysisOptions(
+      // Look for a surrounding "analysis_options.yaml" file.
+      var optionsFile = await findAnalysisOptions(
         _fileSystem,
         await _fileSystem.makePath(file.path),
         resolvePackageUri: (uri) => _resolvePackageUri(file, uri),
       );
 
-      if (options['formatter'] case Map<Object?, Object?> formatter) {
-        if (formatter['page_width'] case int pageWidth) {
-          return _directoryPageWidths[directory] = pageWidth;
+      if (optionsFile['formatter'] case Map<Object?, Object?> formatter) {
+        if (formatter case {'page_width': int width}) {
+          pageWidth = width;
+        }
+
+        if (formatter case {'trailing_commas': String commas}) {
+          trailingCommas = switch (commas) {
+            'automate' => TrailingCommas.automate,
+            'preserve' => TrailingCommas.preserve,
+            // Silently ignore any unrecognized name.
```

Comment:
Should we consider emitting a message on `stderr` for this case?

Reply:
Honestly, I'm not sure. I'm of two minds about this. In general, the formatter tries to never "fail". It just does the best it can with what you give it. It's historically not been the kind of tool to ask for the user's help to run well. So silently ignoring mistakes in configs is consistent with that.

At the same time, it will likely confuse users if they do have a mistake in the analysis_options.yaml file and the formatter just silently does nothing.

Maybe the right way to thread the needle is to print a warning on stderr but continue anyway. I'll think about that some.